OrbitQt/AccessibilityAdapter.cpp

```diff
@@ -0,0 +1,269 @@
+ *                                           |                          |
+ * +-----------+   +----------------------+  |                          |  Adapter for child()
+ * | GlCanvas  +<--+AccessibleInterface   +<----------------------------+
+ * +-----------+   +--+-------------+-----+  |
+ *                    |             ^ Parent |
+ *                    v Child(i)    |        |
+ * +-----------+   +--+-------------+-----+  |                    +-----------------------------+
+ * | TimeGraph +<--+AccessibleInterface   +<----------------------+AccessibilityAdapter         |
+ * +-----------+   +----------------------+  |                    +-----------------------------+
+ *                                           |
+ *                                           +
+ *
+ * The OrbitGlWidgetAccessible is automatically created when an OrbitGlWidget is constructed by
+ * installing a QT Accessibility Factory (InstallAccessibilityFactories()).
+ *
+ * To make sure adapters created by AccessibilityAdapter::GetOrCreateAdapter are deleted when the
+ * corresponding AccessibleInterface is deleted, these interfaces register themselves in the
+ * AccessibleInterfaceRegistry (see AccessibleInterfaceRegistry.h), which in turn allows to register
+ * a callback on interface deletion. AccessibilityAdapter registers itself for this callback and
+ * cleans up interfaces accordingly.
+ */
+
+#include "AccessibilityAdapter.h"
+
+#include <QWidget>
+#include <mutex>
+
+#include "OrbitAccessibility/AccessibleInterfaceRegistry.h"
+
+using orbit_accessibility::AccessibleInterface;
+using orbit_accessibility::AccessibleInterfaceRegistry;
+
+namespace orbit_qt {
+
+absl::flat_hash_map<const AccessibleInterface*, QAccessibleInterface*>
+    AccessibilityAdapter::all_interfaces_map_;
+absl::flat_hash_map<const AccessibleInterface*, std::unique_ptr<AccessibilityAdapter>>
+    AccessibilityAdapter::managed_adapters_;
+
+void AccessibilityAdapter::Init() {
+  AccessibleInterfaceRegistry::Get().SetOnUnregisterCallback(
+      AccessibilityAdapter::OnInterfaceDeleted);
+}
+
+/*
+ * Callback fired when GlAccessibleInterfaces are deleted. This takes care of deleting only those
+ * interfaces created by AccessibilityAdapter.
+ */
+void AccessibilityAdapter::OnInterfaceDeleted(AccessibleInterface* iface) {
+  if (all_interfaces_map_.contains(iface)) {
+    all_interfaces_map_.erase(iface);
+  }
+
+  if (managed_adapters_.contains(iface)) {
+    managed_adapters_.erase(iface);
+  }
+}
+
+QAccessibleInterface* AccessibilityAdapter::GetOrCreateAdapter(const AccessibleInterface* iface) {
+  static std::once_flag flag;
+  std::call_once(flag, Init);
+
+  if (iface == nullptr) {
+    return nullptr;
+  }
+
+  auto it = all_interfaces_map_.find(iface);
+  if (it != all_interfaces_map_.end()) {
+    return it->second;
+  }
+
+  auto adapter = std::unique_ptr<AccessibilityAdapter>(new AccessibilityAdapter(iface));
```

Comment:
`std::make_unique<AccessibilityAdapter>(iface)`

Reply:
Would have been my first choice too, but this doesn't compile because the AccessibilityAdapter constructor is private.

Reply:
Ah right, this `make_unique` with private constructor tricks me every time.